src/coreclr/jit/liveness.cpp

```diff
@@ -1575,7 +1575,11 @@ bool Compiler::fgComputeLifeUntrackedLocal(VARSET_TP&           life,
 
     // We have accurate ref counts when running late liveness so we can eliminate
     // some stores if the lhs local has a ref count of 1.
-    if (isDef && compRationalIRForm && (varDsc.lvRefCnt() == 1) && !varDsc.lvPinned)
+    //
+    // For OSR locals we can't rely on ref counts this way, since some of the appearances
+    // of the local may be in the now-unseen Tier0 portion.
+    //
+    if (isDef && compRationalIRForm && (varDsc.lvRefCnt() == 1) && !varDsc.lvPinned && !varDsc.lvIsOSRLocal)
```

Comment:
Presumably the problem exists only for address-exposed locals (since otherwise we _can_ see all uses, even in the OSR version). Add it as a condition to make the problem more self-evident? (I had to look at the test to understand why this would be illegal.)

Reply:
> Presumably the problem exists only for address-exposed locals (since otherwise we _can_ see all uses, even in the OSR version). Add it as a condition to make the problem more self-evident? (I had to look at the test to understand why this would be illegal.)

Yeah, I can add that.

Another option is to mark any OSR exposed local as implicitly referenced; we do that for pinned locals where we have similar issues not seeing all of the IR.

https://github.com/dotnet/runtime/blob/b5834a0b59047a2157b164fed5da6176527b1949/src/coreclr/jit/lclvars.cpp#L276-L295

Reply:
Using `lvImplicitlyReferenced` seems likely to be a bit more robust so I think I'll change to using that.

Reply:
Does the code in liveness also need to check for `lvImplicitlyReferenced`? Or does something guarantee that `lvRefCount` for implicitly referenced locals is something indeterminately large?

Overall I guess I would assume that all address exposed locals already have to effectively be considered implicitly referenced (whether or not we set that bit), but this seems fine with me.

Reply:
The ref count for implicit locals will be one greater than the number of actual occurrences, so will be at least 2 in this case.